### ISSUE TITLE
The invariant loss do not backforward during training.

### ISSUE THREAD

**Issue by unknowed-ER** [myZeratul/Causal-Debias#2]
Thank you for your outstanding work.
The current code does not seem to calculate gradients when calculating the invariant loss, as it uses `no_ grad` and `detach`. I'm not sure if there's a problem with my understanding. How can the invariant loss participate in gradient backward and update when the Wasserstein distance is non-differentiable?
The code snippet related to the invariant loss is calculated in the `causal-debias.py` file, lines 777-802.
Looking forward to your reply!

```
with torch.no_grad():
    if args.model_type == 'roberta':
        output_invariant = model(input_ids, invariant_mask)
        logits_invariant = output_invariant["logits"]
    else:
        output_invariant = model(input_ids, segment_ids, invariant_mask)
        logits_invariant = output_invariant["logits"]


w_dist = 0

for i in range(logits_invariant.shape[0]):
    w_dist += wasserstein_distance(logits_invariant[i].detach().cpu(),logits_ori[i].detach().cpu())
avg_w_dist = w_dist/(logits_ori.shape[0])

loss_fct = CrossEntropyLoss()


loss = loss_fct(logits_ori.view(-1, num_labels), label_ids.view(-1)) + args.tau * avg_w_dist
```

**Comment by myZeratul** [comment]
The purpose of using "with torch.no_grad()" from line 777 to 783 in the code is to retrieve the distribution predicted by the model for the original sentences corresponding to the augmented ones. This is done to facilitate the calculation of Wasserstein distance later on. Without this context manager, I would need to label the sentences before and after augmentation in my dataset and search for them in subsequent steps. I believe that this approach would increase time complexity, so I opted for this simpler method.

**Comment by unknowed-ER** [comment]
Thank you for your reply.
